raven/utilities/data_assimilation.py

```diff
@@ -0,0 +1,217 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Sun Nov  1 20:48:03 2020
+
+@author: Richard
+"""
+
+
+'''
+model = Raven model instance, preset with parameters etc.
+xa = the set of state variables. In this case, soil0 and soil1 from GR4JCN. Will eventually need to update for other models, other variables
+ts = the timeseries of inputs needed by model (tas, pr, qobs, etc.)
+days = data for the days of assimilation. Assimilation is performed after last day.
+number_members = number of EnKF members
+std = variables for uncertainty estimation of input and streamflow variables.
+    ex: std = {"rainfall": 0.30,"prsn": 0.30, "tasmin": 2.0, "tasmax": 2.0, "water_volume_transport_in_river_channel": 0.15}
+precip = standard deviation used to sample precip, uses gamma distribution (fraction of observed value)
+temperature = standard deviation used to sample temperature, normal dist (degrees Celcius)
+qobs = standard deviation used to sample observed streamflow, normal dist (fraction of observed value)
+'''
+
+import xarray as xr
+import numpy as np
+import tempfile
+from pathlib import Path
+from copy import deepcopy
+
+"""
+Suggestion:
+This was initially suggested to make parallel and faster. However, I was blocked and could not make it work
+I've reverted to coding it in series but it will need to be made parallel as right now it's way too slow (4h for 300 assimilation periods)
```

Comment:
I can get a parallel run to work with a unique "solution" as the starting point (and a state variable for each member). I understand you need to have continuity for each member (a unique solution for each member), correct ? This will need a bit more work to support. Is it preferable to keep the logic as is (loop to have unique solutions), or force a unique solution for each day but run in parallel (with future support for parallel solutions)?

Reply:
Continuity is not really a problem here because the meteo data are perturbed randomly at each step. What the continuity needs to preserve is the link between the final states in "solutions" after the run and the simulated qsim at the end of the run. What goes in the model can be mix-and-matched, but what comes out must be preserved for the assimilation.